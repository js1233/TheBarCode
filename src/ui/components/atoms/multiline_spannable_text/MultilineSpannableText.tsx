import React from "react";
import { StyleProp, Text, TextStyle, View, ViewStyle } from "react-native";
import { AppLabelProps } from "ui/components/atoms/app_label/AppLabel";

export interface SpannableProps {
  containerStyle?: StyleProp<ViewStyle>;
  text: Array<string>;
  appLabelProps: Array<AppLabelProps>;
  numberOfLines?: number;
  rootTextStyle?: StyleProp<TextStyle>;
}

type Props = SpannableProps;

const MultilineSpannableText = React.memo<Props>(
  ({
    text,
    containerStyle,
    appLabelProps,
    rootTextStyle,
    numberOfLines = 0
  }) => {
    if (
      text.length > 0 &&
      appLabelProps.length > 0 &&
      text.length === appLabelProps.length
    ) {
      return (
        <View testID={"SPANNABLE_TEXT"} style={containerStyle}>
          <Text
            numberOfLines={numberOfLines}
            style={[rootTextStyle, { textAlign: "center" }]}>
            {text.map((item, index) => (
              <Text key={index} {...appLabelProps[index]}>
                {item}
              </Text>
            ))}
          </Text>
        </View>
      );
    } else {
      return <View />;
    }
  }
);

export default MultilineSpannableText;
