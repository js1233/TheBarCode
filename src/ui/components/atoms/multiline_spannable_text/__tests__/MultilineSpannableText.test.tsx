import { fireEvent, render } from "@testing-library/react-native";
import React from "react";
import MultilineSpannableText from "ui/components/atoms/multiline_spannable_text/MultilineSpannableText";
import { StyleSheet } from "react-native";

test("check render properly", () => {
  const { getByTestId } = render(
    <MultilineSpannableText
      text={["Male", "Female", "Others"]}
      appLabelProps={[
        { style: styles.firstText },
        { style: styles.email },
        { style: styles.thirdText }
      ]}
    />
  );
  const spannableText = getByTestId("SPANNABLE_TEXT");
  expect(spannableText).toBeDefined();
});

test("Snapshot testing", () => {
  const rendered = render(
    <MultilineSpannableText
      text={["Male", "Female", "Others"]}
      appLabelProps={[
        { style: styles.firstText },
        { style: styles.email },
        { style: styles.thirdText }
      ]}
    />
  ).toJSON();
  expect(rendered).toMatchSnapshot();
});

test("check spannable text on Press", () => {
  const onPress = jest.fn();
  const { getByText } = render(
    <MultilineSpannableText
      text={["Male", "Female", "Others"]}
      appLabelProps={[
        { style: styles.firstText, onPress: () => onPress("Male", 0) },
        { style: styles.email, onPress: () => onPress("Female", 1) },
        { style: styles.thirdText, onPress: () => onPress("Others", 2) }
      ]}
    />
  );
  const spannableText = getByText("Female");
  fireEvent.press(spannableText);
  expect(onPress).toBeCalledWith("Female", 1);
  expect(onPress).toBeCalledTimes(1);
});

const styles = StyleSheet.create({
  firstText: {
    fontWeight: "normal"
  },
  email: {
    fontWeight: "bold"
  },
  thirdText: {
    fontWeight: "normal"
  }
});
