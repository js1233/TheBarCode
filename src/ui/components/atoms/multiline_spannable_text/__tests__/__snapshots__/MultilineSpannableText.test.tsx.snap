// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot testing 1`] = `
<View
  testID="SPANNABLE_TEXT"
>
  <Text
    numberOfLines={0}
    style={
      Array [
        undefined,
        Object {
          "textAlign": "center",
        },
      ]
    }
  >
    <Text
      style={
        Object {
          "fontWeight": "normal",
        }
      }
    >
      Male
    </Text>
    <Text
      style={
        Object {
          "fontWeight": "bold",
        }
      }
    >
      Female
    </Text>
    <Text
      style={
        Object {
          "fontWeight": "normal",
        }
      }
    >
      Others
    </Text>
  </Text>
</View>
`;
